### PR TITLE
PS-7591 Document new variables global.net_buffer_shrink_interval and …

### PR DESCRIPTION
**Pull request by patrickbirch**
…net_buffer_length (8.0)

 Changes to be committed:
	modified:   source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst

### CODE REVIEW

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 1]
```diff
@@ -110,13 +110,34 @@ compiled with ``UNIV_PERF_DEBUG`` C preprocessor define.
 Version Specific Information
 ============================
 
-  * :rn:`8.0.12-1`
-    Feature ported from |Percona Server| 5.7
+  * .. versionadded:: `8.0.12-1` Feature ported from |Percona Server| 5.7
```
Comment (ldonoso) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: It was ported from Facebook MySql 5.6

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 2]
```diff
@@ -110,13 +110,34 @@ compiled with ``UNIV_PERF_DEBUG`` C preprocessor define.
...
+
+ .. _adaptive_network_buffers
+
+ Adaptive Network Buffers 
+ -------------------------
+
+To find the buffer size of the current connection, use the ``network_buffer_length`` status variable. Add ``SHOW GLOBAL`` to review the cumulative buffer sizes for all connections. This variable can help to estimate the maximum size of the network buffers overhead.
+
+Network buffers grows towards the `max_allowed_packet <https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet>`_ size and do not shrink until the connection is terminated. For example, if the connections are selected at random from the pool, an occasional big query eventually increases the buffers of all connections. The combination of `max_allowed packet` set to a value between 64MB to 128MB, and the connection number between 256 to 1024 can create a large memory overhead.
```
Comment (ldonoso) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: A tiny typo, it should be "grow" or "The network buffer of a connection grow ..."

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 3]
```diff
@@ -110,13 +110,34 @@ compiled with ``UNIV_PERF_DEBUG`` C preprocessor define.
...
+
+.. variable:: net_buffer_shrink_interval
+
+   :cmd: --net-buffer-shrink-interval=#
+   :dyn: Yes
+   :scope: Global
+   :vartype: integer 
+   :default: 0
+
+The default value is 0, which disables the variable. The minimum value is 0 and the maximum value is 31536000.
```
Comment (ldonoso) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: I'd use "this functionality" instead of "the variable"

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst (hunk 1 above)
Comment (patrickbirch) [reply doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: Fixed.

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 4]
```diff
@@ -110,13 +110,34 @@ compiled with ``UNIV_PERF_DEBUG`` C preprocessor define.
...
+
+ Adaptive Network Buffers 
+ -------------------------
+
+To find the buffer size of the current connection, use the ``network_buffer_length`` status variable. Add ``SHOW GLOBAL`` to review the cumulative buffer sizes for all connections. This variable can help to estimate the maximum size of the network buffers overhead.
+
+Network buffers grows towards the `max_allowed_packet <https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet>`_ size and do not shrink until the connection is terminated. For example, if the connections are selected at random from the pool, an occasional big query eventually increases the buffers of all connections. The combination of `max_allowed packet` set to a value between 64MB to 128MB, and the connection number between 256 to 1024 can create a large memory overhead.
+
+The :variable:`net_buffer_shrink_interval` variable shrinks network buffers, according to the workload, during the specified interval. The thread's network buffer size is set to the maximum packet size from the last interval. This behavior allows the buffer to forget spikes in the buffer size before the last interval.
```
Comment (ldonoso) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: I'd rephrase this paragraph. Maybe:

>To solve this problem, the variable `net_buffer_shrink_interval` has been introduced. If it is different from 0, percona server records the maximum used size of the network buffer during the number of seconds set by net_buffer_shrink_interval. At the beginning of the next interval the network buffer will be set to this size. This allows to forget spikes in the buffer size

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 5]
```diff
@@ -110,13 +110,34 @@ compiled with ``UNIV_PERF_DEBUG`` C preprocessor define.
...
+
+Network buffers grows towards the `max_allowed_packet <https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet>`_ size and do not shrink until the connection is terminated. For example, if the connections are selected at random from the pool, an occasional big query eventually increases the buffers of all connections. The combination of `max_allowed packet` set to a value between 64MB to 128MB, and the connection number between 256 to 1024 can create a large memory overhead.
+
+The :variable:`net_buffer_shrink_interval` variable shrinks network buffers, according to the workload, during the specified interval. The thread's network buffer size is set to the maximum packet size from the last interval. This behavior allows the buffer to forget spikes in the buffer size before the last interval. 
+
+You can achieve similar results by disconnecting and reconnecting the TCP connections, but this solution is a heavier process. Also, this process also disconnects and reconnects connections with small buffers.
```
Comment (ldonoso) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: I haven't thought of this. This comment is really useful.

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst [hunk 6]
```diff
@@ -110,13 +110,34 @@ compiled with ``UNIV_PERF_DEBUG`` C preprocessor define.
...
+
+Network buffers grows towards the `max_allowed_packet <https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet>`_ size and do not shrink until the connection is terminated. For example, if the connections are selected at random from the pool, an occasional big query eventually increases the buffers of all connections. The combination of `max_allowed packet` set to a value between 64MB to 128MB, and the connection number between 256 to 1024 can create a large memory overhead.
+
+The :variable:`net_buffer_shrink_interval` variable shrinks network buffers, according to the workload, during the specified interval. The thread's network buffer size is set to the maximum packet size from the last interval. This behavior allows the buffer to forget spikes in the buffer size before the last interval. 
+
+You can achieve similar results by disconnecting and reconnecting the TCP connections, but this solution is a heavier process. Also, this process also disconnects and reconnects connections with small buffers. 
+
+.. variable:: net_buffer_shrink_interval
```
Comment (ldonoso) [inline doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst:123]: I would mention the unit of time is seconds. But sorry, I would know where this information should go.

**Review by ldonoso — CHANGES_REQUESTED** [review]

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst (hunk 2 above)
Comment (patrickbirch) [reply doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: Fixed.

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst (hunk 3 above)
Comment (patrickbirch) [reply doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: updated.

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst (hunk 4 above)
Comment (patrickbirch) [reply doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: Rewritten.

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst (hunk 5 above)
Comment (patrickbirch) [reply doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst]: Thanks.

File: doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst (hunk 6 above)
Comment (patrickbirch) [reply doc/source/performance/xtradb_performance_improvements_for_io-bound_highly-concurrent_workloads.rst:123]: Added.

**Review by ldonoso — APPROVED** [review]
